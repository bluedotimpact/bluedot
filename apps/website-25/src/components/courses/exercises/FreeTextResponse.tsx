import CTALinkOrButton from '@bluedot/ui/src/CTALinkOrButton';
import React from 'react';
import ReactMarkdown from 'react-markdown';

type FreeTextResponseProps = {
  // Required
  title: string;
  description: string;
  className?: string;
};

const FreeTextResponse: React.FC<FreeTextResponseProps> = ({
  title,
  description,
  className,
}) => {
  const [submittedOption, setSubmittedOption] = React.useState<string | null>(null);
  const [isSaved, setIsSaved] = React.useState<boolean>(false);

  const handleAnswerSubmit = () => {
    setSubmittedOption('test');
    setIsSaved(true);
  };

  const handleAnswerChange = () => {
    if (submittedOption && isSaved) {
      setIsSaved(false);
    }
  };

  return (
    <div className={`free-text-response container-lined bg-white p-8 flex flex-col gap-6 ${className}`}>
      <div className="free-text-response__header flex flex-col gap-4">
        <div className="free-text-response__header-icon">
          <img src="/icons/lightning_bolt.svg" className="w-15 h-15" alt="" />
        </div>
        <div className="free-text-response__header-content flex flex-col gap-2">
          <p className="free-text-response__title subtitle-sm">{title}</p>
          <ReactMarkdown>{description}</ReactMarkdown>
        </div>
      </div>
      <div className="free-text-response__options flex flex-col gap-2">
        <textarea
          className={`free-text-response__textarea p-4${
            isSaved ? ' free-text-response__textarea--saved container-active bg-[#63C96533] border-[#63C965] text-[#2A5D2A]' : ' container-lined'
          }`}
          placeholder="Enter your answer here"
          onChange={handleAnswerChange}
        />
      </div>
      <CTALinkOrButton
        className="free-text-response__submit"
        variant="primary"
        onClick={handleAnswerSubmit}
        type="button"
      >
        Save
      </CTALinkOrButton>
      {isSaved && <p className="free-text-response__saved-msg">Saved! ðŸŽ‰</p>}
    </div>
  );
};

export default FreeTextResponse;
