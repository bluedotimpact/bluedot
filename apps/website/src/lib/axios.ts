import axios from 'axios';
import { configure } from 'axios-hooks';

// Create axios instance with base URL
const axiosInstance = axios.create({
  baseURL: typeof window !== 'undefined' ? window.location.origin : 'http://localhost:8000',
});

// Configure axios-hooks to use our configured instance
configure({ axios: axiosInstance });

export default axiosInstance;
