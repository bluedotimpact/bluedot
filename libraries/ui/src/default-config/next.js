/**
 * @param {import('next').NextConfig & { headers: Header[] }} [config]
 * @returns {import('next').NextConfig}
 */
const withDefaultBlueDotNextConfig = async (config) => ({
  reactStrictMode: true,
  output: 'standalone',
  distDir: 'dist',
  poweredByHeader: false,
  // We already run eslint as a separate step
  eslint: {
    ignoreDuringBuilds: true,
  },
  devIndicators: false,

  ...config,

  transpilePackages: ['@bluedot/ui', ...(config?.transpilePackages ?? [])],
  headers: async () => {
    return [{
      source: '/:path*',
      headers: [
        {
          key: 'X-BlueDot-Version',
          // eslint-disable-next-line turbo/no-undeclared-env-vars
          value: process.env.VERSION_TAG || 'unknown',
        },
      ],
    },
    {
      source: '/api/:path*',
      headers: [
        {
          key: 'Access-Control-Allow-Origin',
          value: '*',
        },
        {
          key: 'Access-Control-Allow-Methods',
          value: 'GET, POST, PUT, PATCH, DELETE, OPTIONS',
        },
        {
          key: 'Access-Control-Allow-Headers',
          value: 'Content-Type, Authorization',
        },
      ],
    },
    ...(config?.headers ?? [])];
  },
  webpack: (webpackConfig, context, ...args) => {
    // Exclude packages from webpack bundling

    // Only used server side. Trying to import it client-side results in an error
    if (!context.isServer) {
      webpackConfig.externals.push('winston');
    }

    // https://github.com/open-telemetry/opentelemetry-js/issues/4173
    // eslint-disable-next-line no-param-reassign
    webpackConfig.ignoreWarnings = webpackConfig.ignoreWarnings ?? [];
    webpackConfig.ignoreWarnings.push({ module: /opentelemetry/, message: /the request of a dependency is an expression/ });

    // Conditionally reuqired, but we don't actually use it
    webpackConfig.externals.push('@opentelemetry/winston-transport');
    webpackConfig.externals.push('@opentelemetry/exporter-jaeger');

    return config?.webpack?.(config, context, ...args) ?? webpackConfig;
  },
});

module.exports = { withDefaultBlueDotNextConfig };
