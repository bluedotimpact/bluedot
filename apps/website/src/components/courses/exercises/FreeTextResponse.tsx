import clsx from 'clsx';
import { CTALinkOrButton, Tag } from '@bluedot/ui';
import React, { useCallback, useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { useRouter } from 'next/router';
// eslint-disable-next-line import/no-cycle
import MarkdownExtendedRenderer from '../MarkdownExtendedRenderer';
import { getLoginUrl } from '../../../utils/getLoginUrl';
import SaveStatusIndicator from './SaveStatusIndicator';

type SaveStatus = 'idle' | 'typing' | 'saving' | 'saved' | 'error';

type FreeTextResponseProps = {
  className?: string;
  description: string;
  onExerciseSubmit: (exerciseResponse: string, complete?: boolean) => void;
  title: string;
  exerciseResponse?: string;
  isLoggedIn?: boolean;
};

type FormData = {
  answer: string;
};

const FreeTextResponse: React.FC<FreeTextResponseProps> = ({
  className,
  description,
  exerciseResponse,
  isLoggedIn,
  onExerciseSubmit,
  title,
}) => {
  const [saveStatus, setSaveStatus] = useState<SaveStatus>('idle');
  const [lastSavedValue, setLastSavedValue] = useState<string>(exerciseResponse || '');
  const {
    register, handleSubmit, setValue, watch,
  } = useForm<FormData>({
    defaultValues: {
      answer: exerciseResponse || '',
    },
  });
  const router = useRouter();

  // Inject style for Firefox to hide drag notches
  useEffect(() => {
    const styleId = 'free-text-response-firefox-styles';
    if (document.getElementById(styleId)) return;

    const style = document.createElement('style');
    style.id = styleId;
    style.textContent = `
      @-moz-document url-prefix() {
        .drag-notches {
          display: none !important;
        }
      }
    `;
    document.head.appendChild(style);

    // eslint-disable-next-line consistent-return
    return () => {
      const existingStyle = document.getElementById(styleId);
      if (existingStyle) {
        existingStyle.remove();
      }
    };
  }, []);

  useEffect(() => {
    if (exerciseResponse !== undefined) {
      setValue('answer', exerciseResponse);
      setLastSavedValue(exerciseResponse);
    }
  }, [exerciseResponse, setValue]);

  const handleSave = useCallback(async (value: string) => {
    if (!value || value.trim() === lastSavedValue.trim()) {
      setSaveStatus('idle');
      return;
    }

    setSaveStatus('saving');

    try {
      await onExerciseSubmit(value, value.trim().length > 0);

      setLastSavedValue(value);
      setSaveStatus('saved');

      setTimeout(() => {
        setSaveStatus('idle');
      }, 3000);
    } catch (error) {
      setSaveStatus('error');
    }
  }, [onExerciseSubmit, lastSavedValue]);

  // Monitor current value for showing typing status
  const currentValue = watch('answer');
  const hasUserChangedValue = currentValue !== (exerciseResponse || '');
  const isEditing = currentValue !== lastSavedValue;

  useEffect(() => {
    // Show typing status when user has made changes
    if (hasUserChangedValue && isEditing) {
      setSaveStatus('typing');
    }
  }, [currentValue, lastSavedValue, hasUserChangedValue, isEditing]);

  const onSubmit = useCallback(async (data: FormData) => {
    await handleSave(data.answer);
  }, [handleSave]);

  const handleTextareaBlur = useCallback(() => {
    const currentVal = watch('answer') || '';
    // Only auto-save if user has changed value and it's different from saved
    if (isLoggedIn && hasUserChangedValue && currentVal !== lastSavedValue) {
      handleSave(currentVal);
    }
  }, [lastSavedValue, isLoggedIn, hasUserChangedValue, handleSave, watch]);

  const handleRetry = useCallback(() => {
    const currentVal = watch('answer') || '';
    if (currentVal && currentVal !== lastSavedValue) {
      handleSave(currentVal);
    }
  }, [watch, lastSavedValue, handleSave]);

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={clsx('container-lined bg-white p-8 flex flex-col gap-6', className)}>
      <div className="flex flex-col gap-4">
        <Tag variant="secondary" className="uppercase">
          Think it through
        </Tag>
        <div className="flex flex-col gap-2">
          <p className="bluedot-h4 not-prose">{title}</p>
          <MarkdownExtendedRenderer>{description}</MarkdownExtendedRenderer>
        </div>
      </div>
      <div className="flex flex-col relative">
        {/* Wrapper that clips drag notches to textarea boundaries */}
        <div className="relative w-full overflow-hidden rounded-[10px] z-[1]">
          <textarea
            {...register('answer')}
            className={clsx(
              'box-border w-full min-h-[140px] bg-white rounded-[10px] px-6 py-5',
              'font-normal text-[15px] leading-[160%] tracking-[-0.002em] text-[#13132E]',
              'resize-y outline-none transition-all duration-200 block',
              'border-[0.5px] border-[rgba(19,19,46,0.25)]',
              'focus:border-[1.25px] focus:border-[#1641D9] focus:shadow-[0px_0px_10px_rgba(34,68,187,0.3)]',
              'disabled:cursor-not-allowed disabled:opacity-60',
              '[&::-webkit-resizer]:hidden',
            )}
            onBlur={handleTextareaBlur}
            placeholder={isLoggedIn ? 'Enter your answer here' : 'Create an account to save your answers'}
            disabled={!isLoggedIn}
            aria-label={`Writing exercise: ${title}`}
            aria-describedby={isLoggedIn ? 'save-status-message' : undefined}
            aria-required="false"
          />
          {/* Custom drag notches overlay - positioned within the clipping wrapper */}
          <div className="drag-notches absolute w-[15px] h-[14px] right-2 bottom-2 pointer-events-none z-[2]">
            <svg width="15" height="14" viewBox="0 0 15 14" fill="none" xmlns="http://www.w3.org/2000/svg">
              <g opacity="0.6" clipPath="url(#clip0_910_4280)">
                <path d="M11.875 7L7.5 11.375" stroke="#13132E" strokeLinecap="round" strokeLinejoin="round" />
                <path d="M11 2.1875L2.6875 10.5" stroke="#13132E" strokeLinecap="round" strokeLinejoin="round" />
              </g>
              <defs>
                <clipPath id="clip0_910_4280">
                  <rect width="14" height="14" fill="white" transform="translate(0.5)" />
                </clipPath>
              </defs>
            </svg>
          </div>
        </div>
        {isLoggedIn && (
          <SaveStatusIndicator
            status={saveStatus}
            isEditing={isEditing}
            id="save-status-message"
            onRetry={handleRetry}
          />
        )}
      </div>
      {!isLoggedIn && (
        <CTALinkOrButton
          variant="primary"
          url={getLoginUrl(router.asPath, true)}
          withChevron
        >
          Create a free account to save your answers
        </CTALinkOrButton>
      )}
    </form>
  );
};

export default FreeTextResponse;
