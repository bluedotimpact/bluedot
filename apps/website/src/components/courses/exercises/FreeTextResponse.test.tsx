import {
  render, waitFor, fireEvent, getByRole, getByText,
} from '@testing-library/react';
import {
  beforeEach,
  describe,
  expect,
  Mock,
  test,
  vi,
} from 'vitest';
import axios from 'axios';
import { useRouter } from 'next/router';
import FreeTextResponse from './FreeTextResponse';

// Mock next/router
vi.mock('next/router', () => ({
  useRouter: vi.fn(),
}));

// Mock axios
vi.mock('axios');
// Setup axios mock to resolve successfully
(axios.put as Mock).mockResolvedValue({ data: {} });

const mockRouter = {
  asPath: '/test-path',
};

// Setup router mock before each test
beforeEach(() => {
  (useRouter as Mock).mockReturnValue(mockRouter);
});

const mockArgs = {
  title: 'Understanding LLMs',
  description: 'Why is a language model\'s ability to predict \'the next word\' capable of producing complex behaviors like solving maths problems?',
  exerciseId: 'rec1234567890',
  onExerciseSubmit: vi.fn(),
};

describe('FreeTextResponse', () => {
  test('renders default as expected', async () => {
    const { container } = render(
      <FreeTextResponse {...mockArgs} />,
    );

    expect(container).toMatchSnapshot();
    expect(container.querySelector('#save-status-message')).toBeFalsy();
  });

  test('renders logged in as expected', async () => {
    const { container } = render(
      <FreeTextResponse {...mockArgs} isLoggedIn />,
    );

    expect(container).toMatchSnapshot();
    expect(container.querySelector('#save-status-message')).toBeFalsy();
  });

  test('renders with saved exercise response', async () => {
    const { container } = render(
      <FreeTextResponse {...mockArgs} exerciseResponse="This is my saved answer." isLoggedIn />,
    );

    const textareaEl = container.querySelector('textarea') as HTMLTextAreaElement;

    expect(container).toMatchSnapshot();
    expect(textareaEl.value).toBe('This is my saved answer.');
  });

  describe('Auto-save functionality', () => {
    beforeEach(() => {
      vi.clearAllMocks();
      // Reset axios mock to resolve successfully by default
      (axios.put as Mock).mockResolvedValue({ data: {} });
    });

    test('shows typing status when user types in textarea', async () => {
      const { container } = render(
        <FreeTextResponse {...mockArgs} isLoggedIn />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      // Type in the textarea
      fireEvent.change(textarea, { target: { value: 'This is my answer' } });

      // Wait for status to update
      await waitFor(() => {
        const statusMessage = getByText(container, 'Click outside to save your answer');
        expect(statusMessage).toBeTruthy();
      });
    });

    test('triggers auto-save when user clicks outside after typing', async () => {
      const mockOnExerciseSubmit = vi.fn().mockResolvedValue({});
      const { container } = render(
        <FreeTextResponse {...mockArgs} onExerciseSubmit={mockOnExerciseSubmit} isLoggedIn />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      // Type in the textarea
      fireEvent.change(textarea, { target: { value: 'This is my answer' } });

      // Wait for typing status
      await waitFor(() => {
        expect(getByText(container, 'Click outside to save your answer')).toBeTruthy();
      });

      // Trigger blur event (clicking outside)
      fireEvent.blur(textarea);

      // Check that saving status appears
      await waitFor(() => {
        expect(getByText(container, 'Saving answer...')).toBeTruthy();
      });

      // Check that onExerciseSubmit was called
      expect(mockOnExerciseSubmit).toHaveBeenCalledWith('This is my answer', true);

      // Wait for save to complete and show success
      await waitFor(() => {
        expect(getByText(container, 'Answer saved')).toBeTruthy();
      }, { timeout: 1000 });
    });

    test('does not auto-save when user is not logged in', async () => {
      const mockOnExerciseSubmit = vi.fn();
      const { container } = render(
        <FreeTextResponse {...mockArgs} onExerciseSubmit={mockOnExerciseSubmit} isLoggedIn={false} />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      // Type in the textarea
      fireEvent.change(textarea, { target: { value: 'This is my answer' } });

      // Trigger blur event
      fireEvent.blur(textarea);

      // Should not call onExerciseSubmit
      expect(mockOnExerciseSubmit).not.toHaveBeenCalled();

      // Should not show any save status (save status indicator only shows when logged in)
      expect(container.querySelector('#save-status-message')).toBeFalsy();
    });

    test('does not auto-save when content has not changed', async () => {
      const mockOnExerciseSubmit = vi.fn();
      const { container } = render(
        <FreeTextResponse
          {...mockArgs}
          onExerciseSubmit={mockOnExerciseSubmit}
          exerciseResponse="Existing answer"
          isLoggedIn
        />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      // Trigger blur without changing content
      fireEvent.blur(textarea);

      // Should not call onExerciseSubmit
      expect(mockOnExerciseSubmit).not.toHaveBeenCalled();
    });

    test('shows error status when save fails', async () => {
      const mockOnExerciseSubmit = vi.fn().mockRejectedValue(new Error('Save failed'));
      const { container } = render(
        <FreeTextResponse {...mockArgs} onExerciseSubmit={mockOnExerciseSubmit} isLoggedIn />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      // Type in the textarea
      fireEvent.change(textarea, { target: { value: 'This is my answer' } });

      // Trigger blur event
      fireEvent.blur(textarea);

      // Wait for saving status
      await waitFor(() => {
        expect(getByText(container, 'Saving answer...')).toBeTruthy();
      });

      // Wait for error status to appear
      await waitFor(() => {
        expect(getByText(container, "Couldn't save answer.")).toBeTruthy();
        expect(getByRole(container, 'button', { name: /retry/i })).toBeTruthy();
      }, { timeout: 1000 });

      expect(mockOnExerciseSubmit).toHaveBeenCalledWith('This is my answer', true);
    });

    test('retry functionality works when save fails', async () => {
      const mockOnExerciseSubmit = vi.fn()
        .mockRejectedValueOnce(new Error('Save failed'))
        .mockResolvedValueOnce({});

      const { container } = render(
        <FreeTextResponse {...mockArgs} onExerciseSubmit={mockOnExerciseSubmit} isLoggedIn />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      // Type in the textarea
      fireEvent.change(textarea, { target: { value: 'This is my answer' } });

      // Trigger blur event
      fireEvent.blur(textarea);

      // Wait for error status to appear
      await waitFor(() => {
        expect(getByText(container, "Couldn't save answer.")).toBeTruthy();
      }, { timeout: 1000 });

      // Click retry button
      const retryButton = getByRole(container, 'button', { name: /retry/i });
      fireEvent.click(retryButton);

      // Should show saving status again
      await waitFor(() => {
        expect(getByText(container, 'Saving answer...')).toBeTruthy();
      });

      // Should eventually show success
      await waitFor(() => {
        expect(getByText(container, 'Answer saved')).toBeTruthy();
      }, { timeout: 1000 });

      // Should have been called twice (initial attempt + retry)
      expect(mockOnExerciseSubmit).toHaveBeenCalledTimes(2);
      expect(mockOnExerciseSubmit).toHaveBeenCalledWith('This is my answer', true);
    });

    test('success status shows after save completes', async () => {
      const mockOnExerciseSubmit = vi.fn().mockResolvedValue({});
      const { container } = render(
        <FreeTextResponse {...mockArgs} onExerciseSubmit={mockOnExerciseSubmit} isLoggedIn />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      // Type in the textarea
      fireEvent.change(textarea, { target: { value: 'This is my answer' } });

      // Trigger blur event
      fireEvent.blur(textarea);

      // Wait for success status to appear
      await waitFor(() => {
        expect(getByText(container, 'Answer saved')).toBeTruthy();
      }, { timeout: 1000 });

      // Check that the success message is visible
      expect(getByText(container, 'Answer saved')).toBeTruthy();
    });

    test('textarea has proper accessibility attributes', () => {
      const { container } = render(
        <FreeTextResponse {...mockArgs} isLoggedIn />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      expect(textarea.getAttribute('aria-label')).toBe(`Writing exercise: ${mockArgs.title}`);
      expect(textarea.getAttribute('aria-describedby')).toBe('save-status-message');
      expect(textarea.getAttribute('aria-required')).toBe('false');
    });

    test('textarea shows different placeholder when not logged in', () => {
      const { container } = render(
        <FreeTextResponse {...mockArgs} isLoggedIn={false} />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      expect(textarea.placeholder).toBe('Create an account to save your answers');
      expect(textarea.disabled).toBe(true);
    });

    test('textarea shows correct placeholder when logged in', () => {
      const { container } = render(
        <FreeTextResponse {...mockArgs} isLoggedIn />,
      );

      const textarea = container.querySelector('textarea') as HTMLTextAreaElement;

      expect(textarea.placeholder).toBe('Enter your answer here');
      expect(textarea.disabled).toBe(false);
    });
  });
});
