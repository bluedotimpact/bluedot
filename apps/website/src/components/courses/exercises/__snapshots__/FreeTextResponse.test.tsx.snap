// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FreeTextResponse > renders default as expected 1`] = `
<div>
  <form
    class="container-lined bg-white p-8 flex flex-col gap-6"
  >
    <div
      class="flex flex-col gap-4"
    >
      <span
        class="tag inline-flex items-center px-4 py-2 text-xs font-semibold w-fit !text-bluedot-normal bg-[#E5EDFE] rounded-sm uppercase"
        role="status"
      >
        Think it through
      </span>
      <div
        class="flex flex-col gap-2"
      >
        <p
          class="bluedot-h4 not-prose"
        >
          Understanding LLMs
        </p>
        <div
          class="markdown-extended-renderer prose prose-p:text-[16px] prose-li:text-[16px] prose-p:leading-[160%] prose-li:leading-[160%] prose-p:tracking-[-0.002em] prose-li:tracking-[-0.002em] prose-p:font-normal prose-li:font-normal prose-p:font-sans prose-li:font-sans prose-strong:font-sans prose-strong:font-semibold prose-strong:text-[16px] prose-strong:leading-[160%] prose-strong:tracking-[-0.002em] max-w-none"
        />
      </div>
    </div>
    <div
      class="flex flex-col relative"
    >
      <div
        class="relative w-full overflow-hidden rounded-[10px] z-[1]"
      >
        <textarea
          aria-label="Writing exercise: Understanding LLMs"
          aria-required="false"
          class="box-border w-full min-h-[140px] bg-white rounded-[10px] px-6 py-5 font-normal text-[15px] leading-[160%] tracking-[-0.002em] text-[#13132E] resize-y outline-none transition-all duration-200 block border-[0.5px] border-[rgba(19,19,46,0.25)] focus:border-[1.25px] focus:border-[#1641D9] focus:shadow-[0px_0px_10px_rgba(34,68,187,0.3)] disabled:cursor-not-allowed disabled:opacity-60 [&::-webkit-resizer]:hidden"
          disabled=""
          name="answer"
          placeholder="Create an account to save your answers"
        />
        <div
          class="drag-notches absolute w-[15px] h-[14px] right-2 bottom-2 pointer-events-none z-[2]"
        >
          <svg
            fill="none"
            height="14"
            viewBox="0 0 15 14"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_910_4280)"
              opacity="0.6"
            >
              <path
                d="M11.875 7L7.5 11.375"
                stroke="#13132E"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
              <path
                d="M11 2.1875L2.6875 10.5"
                stroke="#13132E"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </g>
            <defs>
              <clippath
                id="clip0_910_4280"
              >
                <rect
                  fill="white"
                  height="14"
                  transform="translate(0.5)"
                  width="14"
                />
              </clippath>
            </defs>
          </svg>
        </div>
      </div>
    </div>
    <a
      class="cta-button flex items-center justify-center transition-all duration-200 w-fit whitespace-nowrap cursor-pointer not-prose text-sm px-4 py-3 rounded-sm font-[650] cta-button--primary bg-bluedot-normal link-on-dark"
      href="http://localhost:3000/login?register=true&redirect_to=%2Ftest-path"
      tabindex="0"
    >
      <span
        class="cta-button__text"
      >
        Create a free account to save your answers
      </span>
      <span
        class="cta-button__chevron ml-3"
      >
        <svg
          class="cta-button__chevron-icon size-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 320 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M310.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-192 192c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L242.7 256 73.4 86.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l192 192z"
          />
        </svg>
      </span>
    </a>
  </form>
</div>
`;

exports[`FreeTextResponse > renders logged in as expected 1`] = `
<div>
  <form
    class="container-lined bg-white p-8 flex flex-col gap-6"
  >
    <div
      class="flex flex-col gap-4"
    >
      <span
        class="tag inline-flex items-center px-4 py-2 text-xs font-semibold w-fit !text-bluedot-normal bg-[#E5EDFE] rounded-sm uppercase"
        role="status"
      >
        Think it through
      </span>
      <div
        class="flex flex-col gap-2"
      >
        <p
          class="bluedot-h4 not-prose"
        >
          Understanding LLMs
        </p>
        <div
          class="markdown-extended-renderer prose prose-p:text-[16px] prose-li:text-[16px] prose-p:leading-[160%] prose-li:leading-[160%] prose-p:tracking-[-0.002em] prose-li:tracking-[-0.002em] prose-p:font-normal prose-li:font-normal prose-p:font-sans prose-li:font-sans prose-strong:font-sans prose-strong:font-semibold prose-strong:text-[16px] prose-strong:leading-[160%] prose-strong:tracking-[-0.002em] max-w-none"
        />
      </div>
    </div>
    <div
      class="flex flex-col relative"
    >
      <div
        class="relative w-full overflow-hidden rounded-[10px] z-[1]"
      >
        <textarea
          aria-describedby="save-status-message"
          aria-label="Writing exercise: Understanding LLMs"
          aria-required="false"
          class="box-border w-full min-h-[140px] bg-white rounded-[10px] px-6 py-5 font-normal text-[15px] leading-[160%] tracking-[-0.002em] text-[#13132E] resize-y outline-none transition-all duration-200 block border-[0.5px] border-[rgba(19,19,46,0.25)] focus:border-[1.25px] focus:border-[#1641D9] focus:shadow-[0px_0px_10px_rgba(34,68,187,0.3)] disabled:cursor-not-allowed disabled:opacity-60 [&::-webkit-resizer]:hidden"
          name="answer"
          placeholder="Enter your answer here"
        />
        <div
          class="drag-notches absolute w-[15px] h-[14px] right-2 bottom-2 pointer-events-none z-[2]"
        >
          <svg
            fill="none"
            height="14"
            viewBox="0 0 15 14"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_910_4280)"
              opacity="0.6"
            >
              <path
                d="M11.875 7L7.5 11.375"
                stroke="#13132E"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
              <path
                d="M11 2.1875L2.6875 10.5"
                stroke="#13132E"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </g>
            <defs>
              <clippath
                id="clip0_910_4280"
              >
                <rect
                  fill="white"
                  height="14"
                  transform="translate(0.5)"
                  width="14"
                />
              </clippath>
            </defs>
          </svg>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`FreeTextResponse > renders with saved exercise response 1`] = `
<div>
  <form
    class="container-lined bg-white p-8 flex flex-col gap-6"
  >
    <div
      class="flex flex-col gap-4"
    >
      <span
        class="tag inline-flex items-center px-4 py-2 text-xs font-semibold w-fit !text-bluedot-normal bg-[#E5EDFE] rounded-sm uppercase"
        role="status"
      >
        Think it through
      </span>
      <div
        class="flex flex-col gap-2"
      >
        <p
          class="bluedot-h4 not-prose"
        >
          Understanding LLMs
        </p>
        <div
          class="markdown-extended-renderer prose prose-p:text-[16px] prose-li:text-[16px] prose-p:leading-[160%] prose-li:leading-[160%] prose-p:tracking-[-0.002em] prose-li:tracking-[-0.002em] prose-p:font-normal prose-li:font-normal prose-p:font-sans prose-li:font-sans prose-strong:font-sans prose-strong:font-semibold prose-strong:text-[16px] prose-strong:leading-[160%] prose-strong:tracking-[-0.002em] max-w-none"
        />
      </div>
    </div>
    <div
      class="flex flex-col relative"
    >
      <div
        class="relative w-full overflow-hidden rounded-[10px] z-[1]"
      >
        <textarea
          aria-describedby="save-status-message"
          aria-label="Writing exercise: Understanding LLMs"
          aria-required="false"
          class="box-border w-full min-h-[140px] bg-white rounded-[10px] px-6 py-5 font-normal text-[15px] leading-[160%] tracking-[-0.002em] text-[#13132E] resize-y outline-none transition-all duration-200 block border-[0.5px] border-[rgba(19,19,46,0.25)] focus:border-[1.25px] focus:border-[#1641D9] focus:shadow-[0px_0px_10px_rgba(34,68,187,0.3)] disabled:cursor-not-allowed disabled:opacity-60 [&::-webkit-resizer]:hidden"
          name="answer"
          placeholder="Enter your answer here"
        />
        <div
          class="drag-notches absolute w-[15px] h-[14px] right-2 bottom-2 pointer-events-none z-[2]"
        >
          <svg
            fill="none"
            height="14"
            viewBox="0 0 15 14"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_910_4280)"
              opacity="0.6"
            >
              <path
                d="M11.875 7L7.5 11.375"
                stroke="#13132E"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
              <path
                d="M11 2.1875L2.6875 10.5"
                stroke="#13132E"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </g>
            <defs>
              <clippath
                id="clip0_910_4280"
              >
                <rect
                  fill="white"
                  height="14"
                  transform="translate(0.5)"
                  width="14"
                />
              </clippath>
            </defs>
          </svg>
        </div>
      </div>
    </div>
  </form>
</div>
`;
